subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyHandler.java

```diff
@@ -203,10 +204,14 @@ private Dependency doAddRegularDependency(Configuration configuration, Object de
 
     @Nullable
     private Dependency doAddProvider(Configuration configuration, Provider<?> dependencyNotation, Closure<?> configureClosure) {
-        if (dependencyNotation instanceof DefaultValueSourceProviderFactory.ValueSourceProvider) {
-            Class<? extends ValueSource<?, ?>> valueSourceType = ((DefaultValueSourceProviderFactory.ValueSourceProvider<?, ?>) dependencyNotation).getValueSourceType();
-            if (valueSourceType.isAssignableFrom(DependencyBundleValueSource.class)) {
-                return doAddListProvider(configuration, dependencyNotation, configureClosure);
+        // See AbstractExternalDependencyFactory$BundleFactory#createBundle, it wraps the base provider with a transformed one
+        if (dependencyNotation instanceof TransformBackedProvider) {
+            Provider<?> baseProvider = ((TransformBackedProvider<?, ?>) dependencyNotation).getProvider();
+            if (baseProvider instanceof DefaultValueSourceProviderFactory.ValueSourceProvider) {
+                Class<? extends ValueSource<?, ?>> valueSourceType = ((DefaultValueSourceProviderFactory.ValueSourceProvider<?, ?>) baseProvider).getValueSourceType();
+                if (valueSourceType.isAssignableFrom(DependencyBundleValueSource.class)) {
+                    return doAddListProvider(configuration, dependencyNotation, configureClosure);
+                }
```

Comment:
I'd consider extracting a new method here to hide all the gory details of the unwrapping:

```
if (isWrappedBundle(dependencyNotation)) {
  return doAddListProvider(configuration, dependencyNotation, configureClosure);
}
```

and then turning the line comment into javadoc.

Reply:
I did this, thanks.